app/validators/well_formed_content_types_validator.rb

```diff
@@ -0,0 +1,78 @@
+class WellFormedContentTypesValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    @error_messages = []
+
+    validate!(value)
+
+    if @error_messages.present?
+      record.errors[attribute] ||= []
+      record.errors[attribute].concat(@error_messages)
+    end
+  end
+
+  private
+
+  def validate!(object)
+    if object.is_a?(Hash)
+      validate_hash!(object)
+    elsif object.is_a?(Array)
+      validate_array!(object)
+    end
+  end
+
+  def validate_hash!(hash)
+    validate_array!(hash.values)
+  end
+
+  def validate_array!(array)
+    if has_content_types?(array)
+      validate_content!(array)
+    else
+      array.all? { |e| validate!(e) }
+    end
+  end
+
+  def has_content_types?(array)
+    return false unless array.all? { |v| v.is_a?(Hash) }
+
+    hash_keys = array.flat_map(&:keys).map(&:to_sym)
+    hash_keys.include?(:content_type)
+  end
```

Comment:
This method is duplicated in the resolver. Ideas for DRYing this up would be appreciated.